src/presentation/cli/controller/runtime.go

```diff
@@ -0,0 +1,240 @@
+package cliController
+
+import (
+	"errors"
+	"log/slog"
+	"strconv"
+	"strings"
+
+	"github.com/speedianet/os/src/domain/valueObject"
+	infraHelper "github.com/speedianet/os/src/infra/helper"
+	internalDbInfra "github.com/speedianet/os/src/infra/internalDatabase"
+	cliHelper "github.com/speedianet/os/src/presentation/cli/helper"
+	"github.com/speedianet/os/src/presentation/service"
+	sharedHelper "github.com/speedianet/os/src/presentation/shared/helper"
+	"github.com/spf13/cobra"
+)
+
+type RuntimeController struct {
+	runtimeService *service.RuntimeService
+}
+
+func NewRuntimeController(
+	persistentDbService *internalDbInfra.PersistentDatabaseService,
+) *RuntimeController {
+	return &RuntimeController{
+		runtimeService: service.NewRuntimeService(persistentDbService),
+	}
+}
+
+func getHostname(hostnameStr string) (hostname valueObject.Fqdn, err error) {
+	primaryVhost, err := infraHelper.GetPrimaryVirtualHost()
+	if err != nil {
+		return hostname, errors.New("PrimaryVirtualHostNotFound")
+	}
+
+	hostname = primaryVhost
+	if hostnameStr != "" {
+		return valueObject.NewFqdn(hostnameStr)
+	}
+
+	return hostname, nil
+}
+
+func (controller *RuntimeController) ReadPhpConfigs() *cobra.Command {
+	var hostnameStr string
+
+	cmd := &cobra.Command{
+		Use:   "get",
+		Short: "GetPhpConfigs",
+		Run: func(cmd *cobra.Command, args []string) {
+			hostname, err := getHostname(hostnameStr)
+			if err != nil {
+				cliHelper.ResponseWrapper(false, err.Error())
+			}
+
+			requestBody := map[string]interface{}{
+				"hostname": hostname.String(),
+			}
+
+			cliHelper.ServiceResponseWrapper(
+				controller.runtimeService.ReadPhpConfigs(requestBody),
+			)
+		},
+	}
+
+	cmd.Flags().StringVarP(&hostnameStr, "hostname", "n", "", "Hostname")
+	return cmd
+}
+
+func (controller *RuntimeController) parseRawPhpModules(
+	rawPhpModules []string,
+) []interface{} {
+	rawModulesMapSlice := []interface{}{}
+	for _, rawModule := range rawPhpModules {
+		rawModuleParts := strings.Split(rawModule, ":")
```

Comment:
Hmmm, that's not how it's done in Control. If you look at the API there, like Container controller and the parsePortBindings, you'll see that since we receive the fields as []interface{} I actually parse them and convert into a slice of the proper valueObject. On the service side I just assert they are of type []valueObject something. That way on the CLI I'll receive them as a string format, say name:value and have a FromString to convert from name:value to the value more close to the DTO when needed. That way I can also use the exact same "serialization" pattern on the infra later to save on the database and read from database easily.

You choose to create a []interface{} of raw strings here and convert into value objects on the service layer so you had to do nothing on the API side. However, parsing a complex object is always something that the API or CLI should do on their own way because only they know which format the sub-level object would come. Here on the CLI it comes as []string{name:value}, on the API it comes as []interface{} or may even come as map[string]interface{} if the user is attempting to send just one value instead of a slice of 1.

I would change this to the format used on Control cause 1) it's able to deal with single value map[string]interface{}; 2) the handling of complex/multi-level objects are done where there are more knowledge about the input; 3) read/save of multi-level simplified and controlled by the domain since converting something to a specific format, say String() would be explicitly explained and you'll avoid doing one way in one place, another way in another place and end up with 5 different ways to factor or cast something to a very common usable format like string.

When you expect on the service that a complex/multi-level object comes as a multi-level thing, you're leaning towards an API layout implicitly rather than a more agnostic approach where we only deal with single level data and leave for the controllers to parse what they know is specific to their format on a global format very close to the DTO itself.

Reply:
Done!